pandas/core/internals/blocks.py

```diff
@@ -637,22 +637,25 @@ def _astype(self, dtype, copy=False, errors='raise', values=None,
             # force the copy here
             if values is None:
 
-                if issubclass(dtype.type,
-                              (compat.text_type, compat.string_types)):
+                if self.is_extension:
```

Comment:
@TomAugspurger are you really really sure this is needed? 
all of this already works with IntegerArray so not t really sure what problem you are trying to solve

Reply:
we already do this is my point

Reply:
Not on master? That hit's https://github.com/pandas-dev/pandas/blob/master/pandas/core/internals/blocks.py#L652, which converts to an ndarray, before ever calling the extension array's astype.

Reply:
At least, my new tests fail on master without these changes. I'm not sure if / how IntegerArray is being handled differently.

Reply:
We don't already "do" this (depending on what you mean with "do" .. :-))

The reason this already works correctly for integer dtypes, is because we are converting the values to an array, and then pass this array to the dtypes array constructor (in case of an EA dtype). 

But as Tom tries to argue, always converting to an ndarray (which is what `get_values` does) is not necessarily a good idea. It might not work (in case of sparse), or can also less efficient (what is actually the case for the integers -> eg when going from Int64 -> Int32, we first convert the Int64 EA to object ndarray to then convert (and infer missing values) again to EA of Int32. This could be done much more efficient).

Reply:
> [about Int EA astyping] This could be done much more efficient

And note that this (the efficient shortcut for going from one int EA to another) is actually implemented in `IntegerArray.astype`. We only never call this in case of `Series(EA).astype(..)` due to the current implementation.

Reply:
i am not arguing that we shouldn’t do this
the implementation is way too convoluted currently

Reply:
do you mean the current code in master or this PR?

Reply:
this PR

there is a ton of code in astype to dispatch to extension types already
this adds yet another branch

Reply:
Can you point to this "ton of code"? I don't see any other dispatch to EAs. 
What is done is the generic conversion to array and then the use of `astype_nansafe` (which has a dispatch to EAs, but for the target EA dtype, not for the calling EA dtype)

Reply:
It's a 2 line change, this extra if condition. Which of those two lines is the convoluted one?

Reply:
hmm, I could swear I add an ``.is_extension`` branch already, that's why this looks weird. In any event this has lots of if/then condition. Please make a note / issue to clean this up.